### PR TITLE
sanoid: fix sanoid.conf generation

### PR DESCRIPTION
**Pull request by ju1m**
###### Motivation for this change
`sanoid` currently generates a bogus `sanoid.conf`: defaults of NixOS' options are put into templates and into datasets (making the use of templates useless).

###### Things done
Simplify the generation of `sanoid.conf`.
As in other modules like `services.postfix.config` I'm also renaming options to use original names because they are 1:1 mapping to sanoid's config.

A relevant discussion, on `#nixos-dev`:

> julm : while I'm at writing a fix to nixos/modules/services/backup/sanoid.nix (which generates a bogus configfile) I'm wondering if there is a rationale for using CamelCase for options which are direct mapping to sanoid options which use snake_case? eg. use_template becomes useTemplate in NixOS. I'd personnaly prefer to stick to the original, but what's the policy?
> rnhmjoj : julm: if it's not too late.. i know at least of the matrix-synapse module, which keeps the original snake_case names. there are also the module.settings options from RFC 0042. personally i don't like having different options styles but keeping the original also has value.
> julm : rnhmjoj: thanks!
> julm : I'll try to justify to keep the original names because it makes the code simpler (no mapping to be done before supplying the attrset to toINI)
> rnhmjoj : julm: if you are doing a 1:1 map of options, yes, i would keep the original names

```
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/backup/sanoid.nix
Comment (lopsided98) [inline nixos/modules/services/backup/sanoid.nix:64]: This is a breaking change that needs to be added to the changelog. My personal opinion is that explicitly defined options should use camel case, but I'm not sure if that is the prevailing sentiment.

Maybe we should just get rid of the explicit options altogether and just make users use `settings`?

File: nixos/modules/services/backup/sanoid.nix
Comment (lopsided98) [inline nixos/modules/services/backup/sanoid.nix:19]: Good catch here. This was changed from using null defaults late in the review cycle for the original PR (https://github.com/NixOS/nixpkgs/pull/72060#issuecomment-583912497) and I didn't consider the side effects.

cc @Infinisil

File: nixos/modules/services/backup/sanoid.nix
Comment (lopsided98) [inline nixos/modules/services/backup/sanoid.nix:180]: Why was this removed?

**Review by lopsided98 — CHANGES_REQUESTED** [review]

File: nixos/modules/services/backup/sanoid.nix
Comment (ju1m) [reply nixos/modules/services/backup/sanoid.nix:64]: Right, I'm adding a changelog entry. When the option is just a mapping to an upstream option I find it confusing to introduce a different name and introduce a ˋdescriptionˋ to say so. However having an explicit option has a great added value when it comes to type checking. Here, the sophisticated type around ˋuse_templateˋ is quite impressive/useful. So I would keep those explicit options, just maybe rename ˋsettingsˋ to ˋextraConfˋ to clarify its purpose.

Comment (lopsided98) [reply nixos/modules/services/backup/sanoid.nix:64]: The `settings` option has specific conventions associated with it, see: https://github.com/NixOS/rfcs/pull/42

File: nixos/modules/services/backup/sanoid.nix
Comment (ju1m) [reply nixos/modules/services/backup/sanoid.nix:180]: So, while fixing the bug I was looking to strip unecessary complexity like using ˋmkMergeˋ and then when submiting this PR I thought that having a module changing its own user-faced option is unusual and confusing, and so far I don't see the added value when one can just merge everything in the call to ˋgenerators.toINIˋ. Sounds right for you?

Comment (ju1m) [reply nixos/modules/services/backup/sanoid.nix:180]: Oh, I'm discovering this new ˋsettingsˋ scheme. So indeed it's unusual but it's not using it that will soon be confusing.

File: nixos/modules/services/backup/sanoid.nix
Comment (lopsided98) [inline nixos/modules/services/backup/sanoid.nix:164]: Sorry for neglecting this PR, but I think this should be done through the dataset/template specific settings option.

**Review by lopsided98 — CHANGES_REQUESTED** [review]

File: nixos/modules/services/backup/sanoid.nix
Comment (ju1m) [reply nixos/modules/services/backup/sanoid.nix]: Indeed, you're right, sorry. Is it better know?

**Review by lopsided98 — APPROVED** [review]
This looks good and I have tested it on my systems.

The only thing I noticed is that the comment on line 13 should probably be removed because this module doesn't have defaults anymore.

We really should backport this PR to 20.03 because without it sanoid is pretty broken. We can't backport it as is because it is a breaking change, but I think backporting just the null defaults should work.

**Comment by ju1m** [comment]
I've separated the changes in two commits so that the fix could be cherry-picked in 20.03.

**Review by lopsided98 — APPROVED** [review]
Thanks for making the backport easier. I think this is ready to merge.

@Infinisil perhaps you can take a look at this since you reviewed the initial sanoid PR?

File: nixos/doc/manual/release-notes/rl-2009.xml
Comment (NinjaTrappeur) [inline nixos/doc/manual/release-notes/rl-2009.xml:134]:
```suggestion
       <literal>services.sanoid.datasets.*.use_template</literal> to match upstream name.
```

File: nixos/modules/services/backup/sanoid.nix
Comment (NinjaTrappeur) [reply nixos/modules/services/backup/sanoid.nix:17]: Agree, when reading the nixos docmentation, `null` can be confusing. I'd rather like having an explicit default.

**Review by picnoir — COMMENTED** [review]
Could you adress the release note typo and rebase this PR?

Not a big fan of the implicit default values, but why not. @lopsided98, you're the maintainer, it's your call.

If you're ok with that, I think this PR is ready to be merged.

**Comment by lopsided98** [comment]
> Not a big fan of the implicit default values, but why not. @lopsided98, you're the maintainer, it's your call.

Is there an alternative that allows templates to work correctly?

**Comment by numinit** [comment]
Didn't realize this was a problem, but fix makes sense to me. +1

**Review by numinit — APPROVED** [review]

**Comment by lopsided98** [comment]
I'm trying to find someone with merge privileges who can take a look at this PR. Could you move the release notes to 21.03 so hopefully it will be ready to merge?

**Comment by ju1m** [comment]
@lopsided98 I've rebased against current `master`.

**Review by lopsided98 — CHANGES_REQUESTED** [review]
I just realized this needs to update the NixOS test with the new option names.

**Comment by ju1m** [comment]
@lopsided98 thanks, I've fixed `nixos/tests/sanoid.nix`.

**Comment by ju1m** [comment]
Rebase and fix merge conflict due to the rename `nixos/doc/manual/release-notes/rl-21{03 => 05}.xml` in cfed3b8b226dfd103f1942c55ffa664178832c29

**Review by doronbehar — APPROVED** [review]

**Review by SuperSandro2000 — COMMENTED** [review]
It would be nice if we could map the old options to the new ones.

**Comment by ju1m** [comment]
@SuperSandro2000 old options are now supported as you requested.

File: nixos/modules/services/backup/sanoid.nix
Comment (SuperSandro2000) [inline nixos/modules/services/backup/sanoid.nix:57]: I was more thinking of https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/monitoring/prometheus/default.nix#L681

Comment (ju1m) [reply nixos/modules/services/backup/sanoid.nix:57]: @SuperSandro2000 AFAICS `mkRenamedOptionModule` does not work for options in `submodule`s, that's why I've used `mkAliasDefinitions` instead.

**Pull request merged**
